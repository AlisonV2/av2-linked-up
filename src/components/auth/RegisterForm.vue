<template>
  <form class="auth-form" @submit.prevent="register">
    <div class="row">
      <app-title>Glad you join us!</app-title>
    </div>
    <div class="form-floating mb-3">
      <input
        type="text"
        class="form-control"
        placeholder="Your name"
        v-model="name"
        required
      />
      <label>Public Name</label>
    </div>
    <div class="form-floating mb-3">
      <input
        type="email"
        class="form-control"
        placeholder="name@example.com"
        v-model="email"
        required
      />
      <label>Email</label>
    </div>
    <div class="form-floating mb-3">
      <input
        type="password"
        class="form-control"
        placeholder="Password"
        v-model="password"
        required
      />
      <label>Password</label>
    </div>
    <div>
      <select class="form-select" v-model="role" required>
        <option disabled value="">You are...</option>
        <option value="artist">An artist</option>
        <option value="client">A (future) client</option>
        <!-- <option value="organizer">An organizer</option> -->
        <!-- <option value="admin">Admin</option> -->
      </select>
    </div>
    <div class="error mb-3 mt-3">{{ error }}</div>
    <div class="btn-center">
      <app-button mode="signup-btn">Sign up</app-button>
    </div>
  </form>
</template>

<script>
import * as Sentry from '@sentry/vue';

/**
 * @exports RegisterForm
 * @type {Component}
 * @requires Sentry
 * @vue-data {string} name - v-model
 * @vue-data {string} email - v-model
 * @vue-data {string} password - v-model
 * @vue-data {string} role - v-model
 * @vue-data {string} error
 * @vue-event register
 */
export default {
  name: 'RegisterForm',
  data() {
    return {
      name: '',
      email: '',
      password: '',
      role: '',
      error: null,
    };
  },
  methods: {
    /**
     * @description Dispatch register store action with form data
     * Redirect to AdminProfile on success
     * @method register
     * @param {object} userData
     * @async
     */
    async register() {
      const userData = {
        name: this.name,
        email: this.email,
        password: this.password,
        role: this.role,
      };
      try {
        await this.$store.dispatch('register', userData);
      } catch (err) {
        this.error = err.message;
        Sentry.captureException(err);
        return;
      }
      this.$router.push({ name: 'AdminProfile' });
    },
  },
};
</script>
