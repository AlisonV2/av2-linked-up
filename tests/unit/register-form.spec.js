import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import RegisterForm from '@/components/auth/RegisterForm.vue';
import AppTitle from '@/components/app/Title.vue';
import AppButton from '@/components/app/Button.vue';

let actions = {
  register: jest.fn(),
};

const store = createStore({
  actions,
});

const $router = {
  push: jest.fn(),
};

const config = {
  data() {
    return {
      name: 'Test',
      email: 'test@test.com',
      password: 'test123',
      role: 'artist',
      error: null,
    };
  },
  global: {
    plugins: [store],
    mocks: {
      $router,
    },
    components: {
      'app-title': AppTitle,
      'app-button': AppButton,
    },
  },
};

/**
 * @module RegisterFormTest
 */
describe('RegisterForm.vue', () => {
  it('RegisterForm Snapshot', () => {
    const component = shallowMount(RegisterForm, config);
    expect(component.element).toMatchSnapshot();
  });

  it('Dispatches "register" action on submit', async () => {
    const component = shallowMount(RegisterForm, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.register).toHaveBeenCalled();
  });

  it('Expects router push to have been called with AdminProfile params', async () => {
    const component = shallowMount(RegisterForm, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect($router.push).toHaveBeenCalledWith({ name: 'AdminProfile' });
  });

  it('Expects error message to be empty', () => {
    const component = shallowMount(RegisterForm, config);
    const error = component.find('.error');
    expect(error.text()).toBeFalsy();
  });

  it('Checks if error message is correctly rendered', () => {
    let data = {
      email: '',
      password: '',
      error: 'Email and password are required',
    };
    const component = shallowMount(RegisterForm, {
      ...config,
      data() {
        return data;
      },
    });
    const error = component.find('.error');
    expect(error.text()).toBe(data.error);
  });

  it('Expects an error to be thrown', async () => {
    actions = {
      register: () => {
        throw new Error('Error');
      },
    };
    const component = shallowMount(RegisterForm, {
      ...config,
      data() {
        return {
          name: '',
          email: '',
          password: '',
          role: '',
          error: null,
        };
      },
    });
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.register).toThrow('Error');
  });
});
