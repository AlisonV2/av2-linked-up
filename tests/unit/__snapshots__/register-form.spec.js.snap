// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm.vue RegisterForm Snapshot 1`] = `
<form
  class="auth-form"
>
  <div
    class="row"
  >
    <div
      class="col-12 text-center mb-4"
    >
      <h2
        class="title"
      >
         Glad you join us ! 
      </h2>
    </div>
  </div>
  <div
    class="form-floating mb-3"
  >
    <input
      class="form-control"
      placeholder="Your name"
      required=""
      type="text"
    />
    <label>
      Public Name
    </label>
  </div>
  <div
    class="form-floating mb-3"
  >
    <input
      class="form-control"
      placeholder="name@example.com"
      required=""
      type="email"
    />
    <label>
      Email
    </label>
  </div>
  <div
    class="form-floating mb-3"
  >
    <input
      class="form-control"
      placeholder="Password"
      required=""
      type="password"
    />
    <label>
      Password
    </label>
  </div>
  <div>
    <select
      class="form-select"
      required=""
    >
      <option
        disabled=""
        value=""
      >
        You are...
      </option>
      <option
        value="artist"
      >
        An artist
      </option>
      <option
        value="client"
      >
        A (future) client
      </option>
    </select>
  </div>
  <div
    class="error mb-3 mt-3"
  />
  <div
    class="btn-center"
  >
    <app-button-stub
      mode="signup-btn"
    />
  </div>
  
  
</form>
`;
